### ISSUE TITLE
The naming rule is not recommended on Java

### ISSUE THREAD

**Issue by a-hisame** [wapatesh/fig#20]
This project conventions method names have end of dollar symbol ($) .

But "Java Language Specification(JLS)" says that `The $ sign should be used only in mechanically generated source code or, rarely, to access pre-existing names on legacy systems.`

So, there are no special reason, you should change that naming rule.
(I think using underscore (_) is good)

JLS: https://docs.oracle.com/javase/specs/jls/se8/html/jls-3.html#jls-3.8

**Comment by LeonLiuY** [comment]
Hi, @a-hisame Thanks for suggestion.

Actually I've already known this idiom. And I've used underscore at the beginning.

The reasons why I change to `$` are:
1. I feel underscore (the shape) is not that evident to make distinction with the default method
2. I don't feel that important to follow the original idiom.

Besides, there is a discussion about the `$` at https://news.ycombinator.com/item?id=9994252

Maybe you'll like it.

**Comment by a-hisame** [comment]
Hi, @liuyang1204  Thank you for your reply.

I read discussion but that discussion don't touch about generated class file and third-party tools.

On Java source code only, I think so too that dollar symbol is no problem and like it more than `_`.

But inner class is named `classname$number` automatically and other third-party tools that using class files may recognize `$` as separator of inner class by mistake.

In minor (internal) project I think this is no problem.
But this is open so this project should follow standard (It is my opition).

If you decide that keep this style, `README.md` should show the reason using `$` symbol not recommended by JLS.

c.f. http://stackoverflow.com/questions/19529603/any-risk-using-a-single-dollar-sign-as-a-java-class-name

**Comment by LeonLiuY** [comment]
I agree that, if the `$` causes any problem for the application, it should be avoided.

As my research, indeed, `$` is used for inner class name, and some identifier inside class, by the compiler.

My conclusion is: `$` will cause problem if:
1. Use it in class name
2. Define member attribute starting with it

Your reference is about single `$` class name, I think it is quite different with Fig's case.

While Fig is using `$` only at the end of a method name, I don't think there is obvious problem.

I agree `$` is something strange in Java, and I'm thinking about how to change. But simply changing to `_` is not what I want.

If `!` is available in Java, I would have definitely used it. I want something that is simple to write, but very evident to warn the API user that you'll change the object using it.

My design decision is mainly based on how to make the most clear & expressive code. As I have already changed a lot of Java idiom, just following Java's tradition is not the way of this project.

Anyway, thank you for your suggestion, it is very helpful!

**Comment by yang-ling** [comment]
I didn't know the convention of `$` before this discussion. Then I did some research.

http://stackoverflow.com/questions/7578155/meanings-of-dollar-sign-in-java-method-descriptor
This site says compiler will create methods with name containing `$`.

http://stackoverflow.com/questions/1987603/is-there-a-convention-when-using-java-rmi-to-use-the-dollar-sign-in-a-variab
This explains the convention of `$`

I think there is no problem if fig users only use standard java toolchain, but they may face unexpected risk if they also use third-party toolchain.

For many developers, `$` is a sign of generated codes. I think we need consider this.

Is there any clear merit of `$`, especially comparing to `_`?

**Comment by LeonLiuY** [comment]
I thought more, now I would like to change `$` to `InPlace`, I think this might be a simple change and solve this issue. How do you think?

According to the suggestions on https://news.ycombinator.com/item?id=10001467

In the future we may need to design immutable & mutable versions.

**Comment by a-hisame** [comment]
I agree that changing `$` to `InPlace`.
(And also, I never use mutable methods because I often use immutable programming style when I use `fig`.)

---

By the way, I researched the language of scala because my memory keeps that the class files genereted scala using `$`.

That results, there are some replacing rule:
- The operator methods such as `+` replace `$plus`.
  If you define the method of `def ++(...)`, the name is replaced `$plus$plus`.
  (The replace rule is not on specifications but implementations shows that rules)
- The methods with default parameters such as `def method(a: String, b: String = "")` replace like
  `Test$Inner1;   a   Ljava/lang/String;   b   method   method$default$2        )Test$Inner1$$$outer`.

I watched chained and connected `$`.

Therefore using `$` only at the end of a method name, I think that `$` perhaps will cause unclear problems.

c.f. Scala Replacing Implementations: https://github.com/scala/scala/blob/v2.11.7/src/library/scala/reflect/NameTransformer.scala
